mit_d3m/__init__.py

```diff
@@ -11,63 +11,98 @@
 import tarfile
 
 import boto3
+import botocore
+import botocore.config
+from funcy import memoize
 
 from mit_d3m.dataset import D3MDS
 from mit_d3m.loaders import get_loader
 from mit_d3m.metrics import METRICS_DICT
+from mit_d3m.utils import contains_files
+
+__all__ = (
+    'DATA_PATH',
+    'BUCKET',
+    'load_d3mds',
+    'load_dataset',
+)
+
 
 DATA_PATH = 'data'
 BUCKET = 'd3m-data-dai'
 
 
-def download_dataset(bucket, dataset, root_dir):
-    client = boto3.client('s3')
+@memoize
+def get_client():
+    config = botocore.config.Config(signature_version=botocore.UNSIGNED)
+    client = boto3.client('s3', config=config)
+    return client
+
+
+def get_dataset_tarfile_path(datapath, dataset):
+    return os.path.join(datapath, '{dataset}.tar.gz'.format(dataset=dataset))
+
 
-    print("Downloading dataset {}".format(dataset))
+def get_dataset_dir(datapath, dataset):
+    return os.path.join(datapath, dataset)
 
-    key = 'datasets/' + dataset + '.tar.gz'
-    filename = root_dir + '.tar.gz'
 
-    print("Getting file {} from S3 bucket {}".format(key, bucket))
+def get_dataset_s3_key(dataset):
+    return 'datasets/{dataset}.tar.gz'.format(dataset=dataset)
+
+
+def download_dataset(bucket, key, filename):
+    print("Downloading dataset from s3://{bucket}".format(bucket=bucket))
+    client = get_client()
     client.download_file(Bucket=bucket, Key=key, Filename=filename)
 
-    shutil.rmtree(root_dir, ignore_errors=True)
 
-    print("Extracting {}".format(filename))
-    with tarfile.open(filename, 'r:gz') as tf:
-        tf.extractall(os.path.dirname(root_dir))
+def extract_dataset(src, dst):
+    print("Extracting {}".format(src))
+    shutil.rmtree(dst, ignore_errors=True)
+    with open(src, 'r') as f:
+        tf = tarfile.TarFile(mode='r:gz', fileobj=f)
+        tf.extractall(dst)
 
 
 def load_d3mds(dataset, root=DATA_PATH, force_download=False):
+    if not os.path.exists(root):
+        os.makedirs(root)
+
     if dataset.endswith('_dataset_TRAIN'):
-        dataset = dataset[:-14]
+        dataset = dataset[:len('_dataset_TRAIN')]
```

Comment:
the idea is to avoid using "magic numbers". it takes python 1e-7 seconds to compute this value which pales in comparison with the dataset download:
```
In [1]: import timeit                                                                                             

In [2]: timeit.timeit('len(\'_dataset_TRAIN\')', number=10000)/10000.                                                    
Out[2]: 1.1375149997547851e-07
```

alternately it could be a module-level constant `DATASET_SUFFIX_LENGTH = 14` or similar but I think what I have is a good tradeoff on readability

Reply:
I agree specially with your last sentence.
Perhaps the ideal would be a constant, but having the `len` there is also very explicit.
Another option would be a `.replace('_dataset_TRAIN')`. Your call.

Reply:
i moved '_dataset_TRAIN' to a module-level constant and left in place the len computation